internal/exec/stages/disks/luks.go

```diff
@@ -341,33 +335,52 @@ func (s *stage) createLuks(config types.Config) error {
 	return nil
 }
 
-func (s *stage) checkLuksDeviceExists(luks types.Luks, keyFilePath string) (bool, error) {
-	devAlias := execUtil.DeviceAlias(*luks.Device)
+func (s *stage) isLuksDevice(device string) bool {
+	devAlias := execUtil.DeviceAlias(device)
 	if _, err := s.Logger.LogCmd(
 		exec.Command(distro.CryptsetupCmd(), "isLuks", "--type", "luks2", devAlias),
-		"checking if %v is a luks device", *luks.Device,
+		"checking if %v is a luks device", device,
 	); err != nil {
 		// isLuks returns exit status 1 if the device is not LUKS
-		return false, nil
+		return false
+	}
+	return true
+}
+
+// Check LUKS device against config and open it.
+func (s *stage) reuseLuksDevice(luks types.Luks, keyFilePath string) error {
+	devAlias := execUtil.DeviceAlias(*luks.Device)
+
+	// don't allow clevis device re-use as we can't guarantee that what
+	// is stored in the header is exactly what would be generated from
+	// the given config
+	if luks.Clevis.IsPresent() {
+		return fmt.Errorf("clevis must not be configured")
+	}
+
+	// ephemeral keyfiles won't match the existing device
+	if util.NilOrEmpty(luks.KeyFile.Source) {
+		return fmt.Errorf("keyfile must be configured")
 	}
 
 	if luks.Label != nil {
 		fsInfo, err := execUtil.GetFilesystemInfo(devAlias, true)
 		if err != nil {
-			return false, fmt.Errorf("retrieving filesystem info: %v", err)
+			return fmt.Errorf("retrieving filesystem info: %v", err)
 		}
 		if fsInfo.Label != *luks.Label {
-			return false, nil
+			return fmt.Errorf("configured label %q doesn't match volume label %q", *luks.Label, fsInfo.Label)
```

Comment:
```suggestion
			return fmt.Errorf("configured label %q doesn't match the expected volume label %q", *luks.Label, fsInfo.Label)
```

Maybe just phrase it like “unit tests” do, to make it sure what is _expected_ and what is the _actual_ thing we’ve read.